CC/Sounder/config.cc

```diff
@@ -126,18 +118,48 @@ Config::Config(const std::string& jsonfile, const std::string& directory,
     cal_tx_gain_.shrink_to_fit();
 
     sample_cal_en_ = tddConf.value("sample_calibrate", false);
+    cal_node_ = tddConf.value("calib_node", "");
     imbalance_cal_en_ = tddConf.value("imbalance_calibrate", false);
     beam_sweep_ = tddConf.value("beamsweep", false);
     beacon_ant_ = tddConf.value("beacon_antenna", 0);
     max_frame_ = tddConf.value("max_frame", 0);
 
+    // Load serials file (loads hub, sdr, and rrh serials)
+    std::string serials_str;
+    auto serials_file_ = tddConf.value("serialsFile", "./files/topology.json");
+    Utils::loadTDDConfig(serials_file_, serials_str);
+    const auto jSerials = json::parse(serials_str, nullptr, true, true);
+    json serialsConf;
+
+    num_cells_ = tddConf.value("cells", 1);
     MLPD_TRACE("Number cells: %zu\n", num_cells_);
+    hub_ids_.resize(num_cells_);
     bs_sdr_ids_.resize(num_cells_);
     n_bs_sdrs_.resize(num_cells_);
     n_bs_antennas_.resize(num_cells_);
     num_bs_sdrs_all_ = 0;
-    for (size_t i = 0u; i < num_cells_; i++) {
-        Utils::loadDevices(bs_sdr_file_.at(i), bs_sdr_ids_.at(i));
+
+    reciprocal_calib_ = tddConf.value("internal_measurement", false);
+    for (size_t i = 0; i < num_cells_; i++) {
+        std::string cell_str = "Cell" + std::to_string(i);
+        ss << jSerials[0].value(cell_str, serialsConf);
+        serialsConf = json::parse(ss);
+        ss.str(std::string());
+        ss.clear();
+
+        hub_ids_.at(i) = serialsConf.value("hub", "");
+        auto sdr_serials = serialsConf.value("sdr", json::array());
+        bs_sdr_ids_.at(i).assign(sdr_serials.begin(), sdr_serials.end());
+
+        // Remove cal node if not doing reciprocal calibration
+        if (!reciprocal_calib_) {
+            auto iterCal = std::find(bs_sdr_ids_.at(i).begin(), bs_sdr_ids_.at(i).end(), cal_node_);
```

Comment:
Why do we need a cal_node when `!reciprocal_calib_`?

Reply:
pyfaros will automatically pull all boards in the discovered chains, including any calibration nodes. Since we plan to have a calibration node in each base station we must somehow remove them during normal operation. It will only keep it listed if reciprocal_calib_ is enabled.